clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -200,6 +200,99 @@ bool Sema::isKnownGoodSYCLDecl(const Decl *D) {
   return false;
 }
 
+bool isZeroSizedArray(QualType Ty) {
+  if (const auto *CATy = dyn_cast<ConstantArrayType>(Ty))
+    return CATy->getSize() == 0;
+  return false;
+}
+
+Sema::DeviceDiagBuilder emitDeferredDiagnosticAndNote(Sema &S, SourceRange Loc,
+                                                      unsigned DiagID,
+                                                      SourceRange UsedAtLoc) {
+  Sema::DeviceDiagBuilder builder =
+      S.SYCLDiagIfDeviceCode(Loc.getBegin(), DiagID);
+  if (UsedAtLoc.isValid())
+    S.SYCLDiagIfDeviceCode(UsedAtLoc.getBegin(), diag::note_sycl_used_here);
+  return builder;
+}
+
+void checkSYCLVarType(Sema &S, QualType Ty, SourceRange Loc,
+                      llvm::DenseSet<QualType> Visited,
+                      SourceRange UsedAtLoc = SourceRange()) {
+  // not all variable types are supported in kernel contexts
```

Comment:
This comment needs to use complete sentences and punctuation.  Its not clear ot me what you're saying here.